python/mxnet/amp/amp.py

```diff
@@ -0,0 +1,303 @@
+        except AttributeError:
+            pass
+
+    wrap_list = fp32_list if fp32_list is not None else lists.symbol.FP32_FUNCS
+    for fun_name in wrap_list:
+        try:
+            f_to_wrap = getattr(module, fun_name)
+            setattr(module, fun_name, _wrapper(f_to_wrap, np.float32))
+        except AttributeError:
+            pass
+
+    wrap_list = conditional_fp32_list if conditional_fp32_list is not None \
+                    else lists.symbol.CONDITIONAL_FP32_FUNCS
+    for fun_name, arg, arg_values in wrap_list:
+        try:
+            f_to_wrap = getattr(module, fun_name)
+            setattr(module, fun_name, _wrapper(f_to_wrap, np.float32, (arg, arg_values)))
+        except AttributeError:
+            pass
+
+    for fun_name in lists.symbol.WIDEST_TYPE_CASTS:
+        try:
+            f_to_wrap = getattr(module, fun_name)
+            setattr(module, fun_name, _symbol_widest_wrapper(f_to_wrap))
+        except AttributeError:
+            pass
+
+def _wrap_loss_output_functions(module, ls):
+    if module == ndarray:
+        def _wrapper(f):
+            def _scaling_wrapper(*args, **kwargs):
+                if 'grad_scale' in kwargs:
+                    kwargs['grad_scale'] = kwargs['grad_scale'] * ls.loss_scale
+                else:
+                    kwargs['grad_scale'] = ls.loss_scale
+                return f(*args, **kwargs)
+            _scaling_wrapper.__name__ = f.__name__
+            _scaling_wrapper.__module__ = f.__module__
+            _scaling_wrapper.__doc__ = f.__doc__
+            return _scaling_wrapper
+    else:
+        def _wrapper(f):
+            def _warning_wrapper(*args, **kwargs):
+                logging.warning("%s does not support dynamic loss scaling "
+                                "in symbolic and hybridized execution.", f.__name__)
+                return f(*args, **kwargs)
+            _warning_wrapper.__name__ = f.__name__
+            _warning_wrapper.__module__ = f.__module__
+            _warning_wrapper.__doc__ = f.__doc__
+            return _warning_wrapper
+
+    for fun_name in lists.symbol.LOSS_OUTPUT_FUNCTIONS:
+        try:
+            f_to_wrap = getattr(module, fun_name)
+            setattr(module, fun_name, _wrapper(f_to_wrap))
+        except AttributeError:
+            pass
+
+_amp_initialized = False
+_amp_loss_scale_initialized = False
+_loss_scaler = None
+
+@contextlib.contextmanager
+def scale_loss(loss, optimizer_or_trainer):
+    assert optimizer_or_trainer._amp_loss_scaler is not None, \
+        'Loss scaler is not initialized, did you forget to call amp.init_trainer()?'
+    optimizer_or_trainer._scale = 1. / optimizer_or_trainer._amp_loss_scaler.loss_scale
+    if isinstance(loss, (list, tuple)):
+        yield [l * optimizer_or_trainer._amp_loss_scaler.loss_scale for l in loss]
+    else:
+        yield optimizer_or_trainer._amp_loss_scaler.loss_scale * loss
+
+def init(fp16_list=None, conditional_fp32_list=None, fp32_list=None):
```

Comment:
Thanks for adding this support !
can we change the interface to look something like this : 
```
def init(target_dtype="float16", original_dtype="float32", target_precision_ops=None, original_precision_ops=None, conditional_original_precision_ops=None)
```

This way we can add bf16 support for target_dtype and also add support for other high precision dtypes. WDYT?

Also is there any reason to not include the widest_type_list ?